celestia/src/common.rs

```diff
@@ -0,0 +1,126 @@
+use std::net::SocketAddr;
+use std::path::PathBuf;
+
+use anyhow::{bail, Result};
+use celestia_node::network::Network;
+use clap::{error::ErrorKind, ArgGroup, CommandFactory, Parser, ValueEnum};
+use libp2p::multiaddr::Protocol;
+use libp2p::Multiaddr;
+use serde_repr::{Deserialize_repr, Serialize_repr};
+
+use crate::{native, server};
+
+const SERVER_DEFAULT_BIND_ADDR: &str = "127.0.0.1:9876";
+
+#[derive(Debug, Parser)]
+// disallow specifying store path when running in browser
+#[clap(group(ArgGroup::new("native_xor_browser")
+             .args(&["store", "browser"])))]
+pub(crate) struct Args {
+    /// Network to connect.
+    #[arg(short, long, value_enum, default_value_t)]
+    pub(crate) network: ArgNetwork,
+
+    /// Listening addresses. Can be used multiple times.
+    #[arg(short, long = "listen")]
+    pub(crate) listen_addrs: Vec<Multiaddr>,
+
+    /// Bootnode multiaddr, including peer id. Can be used multiple times.
+    #[arg(short, long = "bootnode")]
+    pub(crate) bootnodes: Vec<Multiaddr>,
+
+    /// Persistent header store path.
+    #[arg(short, long = "store")]
+    pub(crate) store: Option<PathBuf>,
+
+    /// Serve wasm node which can be accessed with web browser
+    #[arg(long)]
+    browser: bool,
+}
+
+#[derive(
+    Debug, Default, Clone, Copy, PartialEq, Eq, ValueEnum, Serialize_repr, Deserialize_repr,
+)]
+#[repr(u8)]
+pub(crate) enum ArgNetwork {
+    Arabica,
+    Mocha,
+    #[default]
+    Private,
+}
+
+pub async fn run_cli() -> Result<()> {
+    let _ = dotenvy::dotenv();
+    let mut args = Args::parse();
+    let _guard = init_tracing();
+
+    if args.browser {
+        let listen_addr = match args.listen_addrs.len() {
+            0 => SERVER_DEFAULT_BIND_ADDR.parse().unwrap(),
+            1 => multiaddr_to_socketaddr(args.listen_addrs.pop().unwrap())?,
+            _ => Args::command()
+                .error(
+                    ErrorKind::TooManyValues,
+                    "expected single listenting address when serving wasm node",
+                )
+                .exit(),
+        };
+        server::run(args.network, args.bootnodes, listen_addr).await
```

Comment:
tbh I'm not a big fan of this and multiaddr_to_socketaddr parsing. What would you say for subcommands? It's fine for me for duplicating some arguments but I think there is also a way to pass common ones.

```rs
#[derive(Parser)]
#[command(author, version, about, long_about = None)]
enum TestSuite {
    Node(NodeArgs),
    Browser(BrowserArgs),
}

#[derive(Args)]
struct NodeArgs {
    ...
}
```

It could then be used as
```bash
celestia node -n mocha
celestia browser -n mocha
```

and you can then have `listen_addr` as `Vec<Multiaddress>` for node and `SocketAddrV4` for browser

Reply:
In general I mostly use subcommands. It is much easier to split the actions. Duplication is fine is this case.

Reply:
I thought simple flag would do, but probably it's good idea to switch to subcommands here, changed 👍